CRYOSMSApp/src/CRYOSMSDriver.cpp

```diff
@@ -109,6 +111,46 @@ asynStatus CRYOSMSDriver::writeInt32(asynUser *pasynUser, epicsInt32 value)
 		setupRamp();
 		return putDb("START:SP", &falseVal);
 	}
+	else if (function == P_calcHeater) {
+		// Deconstructing the "HEATER" command is a little too complicated for streamDevice so we do it here. There are 2 types of readback:
+		// ........ HEATER STATUS: {OFF|ON}
+		// ........ HEATER STATUS: OFF AT {value} {units}
+		// Due to string size constraints we get stream device to cut off everything before the OFF/ON
+		asynStatus status = asynSuccess;
+		std::string heater_resp;
+		// Extract from db
+		RETURN_IF_ASYNERROR2(getDb, "HEATER:STAT:RAW", heater_resp);
+		// Handle simple on/off case
+		if (heater_resp.find("ON") != std::string::npos) {
+			return putDb("HEATER:STAT", &trueVal);
+		}
+		else if (heater_resp.find("OFF") != std::string::npos && heater_resp.find("AT") == std::string::npos)
+		{
+			return putDb("HEATER:STAT", &falseVal);
+		}
+		// If we find "AT"  then we know to expect numbers
+		else if (heater_resp.find("OFF") != std::string::npos && heater_resp.find("AT") != std::string::npos) {
+			RETURN_IF_ASYNERROR2(putDb, "HEATER:STAT", &falseVal);
+			// Check units. Can only be TESLA or AMPS so we just use if / else instead of if / else if / else
+			if (heater_resp.find("TESLA") != std::string::npos) {
+				RETURN_IF_ASYNERROR2(putDb, "OUTPUT:PERSIST:RAW:UNIT", &trueVal);
+			}
+			else {
+				RETURN_IF_ASYNERROR2(putDb, "OUTPUT:PERSIST:RAW:UNIT", &falseVal);
+			}
+			// Find first  and last digits of value
+			int firstNum = heater_resp.find_first_of("1234567890");
+			int lastNum = heater_resp.find_last_of("1234567890");
+			// And snip the middle to another string
+			std::string persistStr(heater_resp, firstNum, lastNum);
```

Comment:
This might be slightly more clear
```suggestion
			std::string persistStr = heater_resp.substr(firstNum, lastNum)
```